tests/core/test_results.py

```diff
@@ -240,7 +230,90 @@ def test_result_gather_mixed_types():
 
 def test_result_retrieve_last_logged_item():
     result = Result()
-    result.log('a', 5., on_step=True, on_epoch=True)
-    assert result['epoch_a'] == 5.
-    assert result['step_a'] == 5.
-    assert result['a'] == 5.
+    result.log("a", 5.0, on_step=True, on_epoch=True)
+    assert result["epoch_a"] == 5.0
+    assert result["step_a"] == 5.0
+    assert result["a"] == 5.0
+
+
+def test_result_train_epoch_end(tmpdir):
+    class LitModel(LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.l1 = torch.nn.Linear(28 * 28, 10)
+
+        def forward(self, x):
+            return torch.relu(self.l1(x.view(x.size(0), -1)))
+
+        def training_step(self, batch, batch_idx):
+            x, y = batch
+            y_hat = self(x)
+            loss = F.cross_entropy(y_hat, y)
+            result = TrainResult(minimize=loss)
+            result.log_dict({"train_loss": loss})
+
+            result.y = y
+            result.y_hat = y_hat
+            return result
+
+        def training_epoch_end(self, out):
+            epoch_loss = F.cross_entropy(out.y_hat, out.y)
+            result = TrainResult()
+            result.log_dict({"train_epoch_loss": epoch_loss})
+            return result
+
+        def configure_optimizers(self):
+            return torch.optim.Adam(self.parameters(), lr=0.0005)
+
+    dm = TrialMNISTDataModule(tmpdir)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        weights_summary=None,
+    )
+    trainer.fit(LitModel(), dm)
+
+
+def test_result_valid_epoch_end(tmpdir):
+    class LitModel(LightningModule):
```

Comment:
How this differs from our `EvalModelTemplate`, pls write just the methods which differ...